src/Compiler/Service/FSharpCheckerResults.fs

```diff
@@ -937,6 +932,85 @@ type internal TypeCheckInfo
 
     let DefaultCompletionItem item = CompletionItem ValueNone ValueNone item
 
+    let CompletionItemSuggestedName displayName =
+        {
+            ItemWithInst = ItemWithNoInst(Item.NewDef(Ident(displayName, range0)))
+            MinorPriority = 0
+            Type = None
+            Kind = CompletionItemKind.SuggestedName
+            IsOwnMember = false
+            Unresolved = None
+        }
+
+    /// Checks whether the suggested name is unused.
+    /// In the future we could use an increasing numeric suffix for conflict resolution
+    let CreateCompletionItemForSuggestedPatternName (pos: pos) name =
+        if String.IsNullOrWhiteSpace name then
+            None
+        else
+            let name = String.lowerCaseFirstChar name
+
+            let unused =
+                sResolutions.CapturedNameResolutions
+                |> ResizeArray.forall (fun r ->
+                    match r.Item with
+                    | Item.Value vref when r.Pos.Line = pos.Line -> vref.DisplayName <> name
+                    | _ -> true)
+
+            if unused then
+                Some(CompletionItemSuggestedName name)
+            else
+                None
+
+    /// Suggest name based on type
+    let SuggestNameBasedOnType g pos ty =
+        if isNumericType g ty then
+            CreateCompletionItemForSuggestedPatternName pos "num"
+        else
+            match tryTcrefOfAppTy g ty with
+            | ValueSome tcref when not (tyconRefEq g g.system_Object_tcref tcref) ->
+                CreateCompletionItemForSuggestedPatternName pos tcref.DisplayName
```

Comment:
(styling-only comment):

When pattern matching on a DU with named fields, it's more future proof to pattern match using the names and not just positions. (in case the DU case gets extended, that particular line is not affected).

What is the opinion on suggesting $"{tcref.DisplayName}={tcref.DisplayName}" ?

Reply:
I see a couple of problems with that.

- It would only work while completing the very first indentifier in the pattern (in which case we should suggest all fields in the named form) or on the last named identifier (then we'd have to look for all the fields that are already matched and filter those out). Sounds a bit complicated to implement at first thought.
- We would no longer be completing an identifier/keyword but a more structured piece of code. Don't think there's a precedent for that yet.

I reckon a code fix/snippet like https://twitter.com/auduchinok/status/1460918204885254146 would be more appropriate for that kind of thing.